controllers/clusterversion_controller.go

```diff
@@ -139,6 +141,32 @@ func (c *ClusterVersionReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, err
 	}
 
+	// deletion phase
+	if !instance.GetDeletionTimestamp().IsZero() {
+		result, err := c.deletionPhase()
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if !result.Requeue {
+			//remove finalizer
+			controllerutil.RemoveFinalizer(&instance, ClusterVersionFinalizer)
+			if err := c.Client.Update(c.ctx, &instance); err != nil {
+				return ctrl.Result{}, err
+			}
+			c.log.Info("finallizer removed successfully")
+		}
+		return result, nil
```

Comment:
There are 3 conditions where we should not remove the finalizer:
1. an error was returned
2. The result indicates the need for a requeue via the `Requeue` field
3. The result indicated the need for a requeue via the `RequeueAfter` field


```suggestion
		if err != nil || !result.Requeue || result.RequeueAfter > 0 {
			return result, err
		}
			
		//remove finalizer
		controllerutil.RemoveFinalizer(&instance, ClusterVersionFinalizer)
		if err := c.Client.Update(c.ctx, &instance); err != nil {
			return ctrl.Result{}, err
		}
		c.log.Info("finallizer removed successfully")
		
		return ctrl.Result{}, nil
		
```